SpatialGDK/Source/SpatialGDKTests/SpatialGDK/EngineClasses/GridBasedLBStrategy/GridBasedLBStrategyTest.cpp

```diff
@@ -0,0 +1,283 @@
+{
+	uint32 Actual = Strat->WhoShouldHaveAuthority(*TestActors[Handle]);
+
+	Test->TestEqual(FString::Printf(TEXT("Who Should Have Authority. Actual: %d, Expected: %d"), Actual, ExpectedVirtualWorker), Actual, ExpectedVirtualWorker);
+
+	return true;
+}
+
+DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FCheckVirtualWorkersDiffer, FAutomationTestBase*, Test, TArray<FName>, Handles);
+bool FCheckVirtualWorkersDiffer::Update()
+{
+	TMap<uint32, FName> WorkerIdToHandle;
+	for (int i = 0; i < Handles.Num(); i++)
+	{
+		const uint32 WorkerId = Strat->WhoShouldHaveAuthority(*TestActors[Handles[i]]);
+		if (WorkerIdToHandle.Contains(WorkerId))
+		{
+			Test->AddError(FString::Printf(TEXT("%s and %s both belong to virtual worker %d"),
+				*WorkerIdToHandle[WorkerId].ToString(), *Handles[i].ToString(), WorkerId));
+		}
+		else
+		{
+			WorkerIdToHandle.Add(WorkerId, Handles[i]);
+		}
+	}
+
+	return true;
+}
+
+DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FCheckVirtualWorkersMatch, FAutomationTestBase*, Test, TArray<FName>, Handles);
+bool FCheckVirtualWorkersMatch::Update()
+{
+	uint32 VirtualWorkerId = Strat->WhoShouldHaveAuthority(*TestActors[Handles[0]]);
+
+	for (int i = 1; i < Handles.Num(); i++)
+	{
+		uint32 NextVirtualWorkerId = Strat->WhoShouldHaveAuthority(*TestActors[Handles[i]]);
+		Test->TestEqual(FString::Printf(TEXT("Should Have Authority %s(%d) and %s(%d)"),
+			*Handles[0].ToString(), VirtualWorkerId, *Handles[i].ToString(), NextVirtualWorkerId),
+			VirtualWorkerId, NextVirtualWorkerId);
+	}
+
+	return true;
+}
+
+GRIDBASEDLBSTRATEGY_TEST(GIVEN_2_rows_3_cols_WHEN_get_virtual_worker_ids_is_called_THEN_it_returns_6_ids)
+{
+	Strat = UTestGridBasedLBStrategy::Create(2, 3, 10000.f, 10000.f);
+	Strat->Init(nullptr);
+
+	TArray<uint32> VirtualWorkerIds = Strat->GetVirtualWorkerIds();
+	TestEqual("Number of Virtual Workers", VirtualWorkerIds.Num(), 6);
+
+	return true;
+}
+
+GRIDBASEDLBSTRATEGY_TEST(GIVEN_a_grid_WHEN_get_virtual_worker_ids_THEN_all_worker_ids_are_valid)
+{
+	Strat = UTestGridBasedLBStrategy::Create(5, 10, 10000.f, 10000.f);
+	Strat->Init(nullptr);
+
+	TArray<uint32> VirtualWorkerIds = Strat->GetVirtualWorkerIds();
+	for (uint32 VirtualWorkerId : VirtualWorkerIds)
+	{
+		TestNotEqual("Virtual Worker Id", VirtualWorkerId, SpatialConstants::INVALID_VIRTUAL_WORKER_ID);
+	}
+
+	return true;
+}
+
+GRIDBASEDLBSTRATEGY_TEST(GIVEN_grid_is_not_ready_WHEN_local_virtual_worker_id_is_set_THEN_is_ready)
+{
+	Strat = UTestGridBasedLBStrategy::Create(1, 1, 10000.f, 10000.f);
+	Strat->Init(nullptr);
+
+	TestFalse("IsReady Before LocalVirtualWorkerId Set", Strat->IsReady());
+
+	Strat->SetLocalVirtualWorkerId(123);
+
+	TestTrue("IsReady After LocalVirtualWorkerId Set", Strat->IsReady());
+
+	return true;
+}
+
+GRIDBASEDLBSTRATEGY_TEST(GIVEN_a_single_cell_and_local_id_1_WHEN_should_relinquish_called_THEN_returns_false)
+{
+	AutomationOpenMap("/Engine/Maps/Entry");
+
+	ADD_LATENT_AUTOMATION_COMMAND(FCreateStrategy(1, 1, 10000.f, 10000.f, 0));
```

Comment:
isn't the local id 0 here?

Reply:
The 0 here actually means the 0th element in the GetVirtualWorkerIds() array, which is 1 in this case.

Reply:
That said, I've tried to avoid having the layout of Ids be part of the contract for the strategy, so have renamed the test.